File: client/src/api/boards.ts
import { api } from './client';
import type { Column } from '../components/KanbanColumn/KanbanColumn';

export interface BoardCreate {
  title: string;
}

export interface BoardUpdate {
  title: string;
}

export interface BoardOut {
  id: number;
  title: string;
  owner_id?: number;
}

export async function listBoards() {
  try {
    const response = await api.get<BoardOut[]>('/boards');
    return response;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to fetch boards');
  }
}

export async function createBoard(data: BoardCreate) {
  try {
    const response = await api.post<BoardOut>('/boards', data);
    return response;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to create board');
  }
}

export async function updateBoard(boardId: number, data: BoardUpdate) {
  try {
    const response = await api.put<BoardOut>(`/boards/${boardId}`, data);
    return response;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to update board');
  }
}

export async function deleteBoard(boardId: number) {
  try {
    const response = await api.delete(`/boards/${boardId}`);
    return response;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to delete board');
  }
}

export async function fetchColumns(boardId: number) {
  try {
    const response = await api.get<Column[]>(`/boards/${boardId}/columns`);
    return response;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to fetch columns');
  }
}

export async function createColumn(boardId: number, title: string) {
  try {
    await api.post<Column>(`/boards/${boardId}/columns`, { title });
    const updatedColumns = await fetchColumns(boardId);
    return updatedColumns;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to create column');
  }
}

export async function updateColumn(boardId: number, columnId: number, title: string) {
  try {
    await api.put<Column>(`/boards/${boardId}/columns/${columnId}`, { title });
    const updatedColumns = await fetchColumns(boardId);
    return updatedColumns;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to update column');
  }
}

export async function deleteColumn(boardId: number, columnId: number) {
  try {
    await api.delete(`/boards/${boardId}/columns/${columnId}`);
    const updatedColumns = await fetchColumns(boardId);
    return updatedColumns;
  } catch (error: any) {
    throw new Error(error.response?.data?.detail || 'Failed to delete column');
  }
}

File: server/app/routers/boards.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from app.database import get_db
from app.models import Board, BoardMember, User
from app.schemas.board_schema import BoardCreate, BoardOut, BoardUpdate
from app.routers.auth import get_current_user

router = APIRouter(prefix="/boards", tags=["boards"])

# ✅ CREATE
@router.post("/", response_model=BoardOut, status_code=status.HTTP_201_CREATED)
def create_board(
    b_in: BoardCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    board = Board(title=b_in.title, owner_id=current_user.id)
    db.add(board)
    db.commit()
    db.refresh(board)

    # เพิ่มตัวสร้างเป็นสมาชิกบอร์ดด้วย
    member = BoardMember(board_id=board.id, user_id=current_user.id)
    db.add(member)
    db.commit()

    return board

# ✅ READ (List all boards of current user)
@router.get("/", response_model=List[BoardOut])
def list_boards(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return (
        db.query(Board)
        .join(BoardMember, BoardMember.board_id == Board.id)
        .filter(BoardMember.user_id == current_user.id)
        .all()
    )

# ✅ READ (Single board)
@router.get("/{board_id}", response_model=BoardOut)
def get_board(
    board_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    board = (
        db.query(Board)
        .join(BoardMember, BoardMember.board_id == Board.id)
        .filter(Board.id == board_id, BoardMember.user_id == current_user.id)
        .first()
    )
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    return board

# ✅ UPDATE
@router.put("/{board_id}", response_model=BoardOut)
def update_board(
    board_id: int,
    b_update: BoardUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    board = db.query(Board).filter(Board.id == board_id).first()
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    if board.owner_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized to update this board")

    board.title = b_update.title
    db.commit()
    db.refresh(board)
    return board

# ✅ DELETE
@router.delete("/{board_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_board(
    board_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    board = db.query(Board).filter(Board.id == board_id).first()
    if not board:
        raise HTTPException(status_code=404, detail="Board not found")
    if board.owner_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized to delete this board")

    db.delete(board)
    db.commit()
